File: leetcode/binary_search/644_Maximum_Average_Subarray_II/codes/convex_hull.c
#include <stdio.h>
#include <stdlib.h>
#include <float.h>

double max(double a, double b) {
    return a > b ? a : b;
}

int is_concave(int x1, int x2, int x3, double* prefix) {
    int dx1 = x2 - x1;
    double dy1 = prefix[x2] - prefix[x1];
    int dx2 = x3 - x2;
    double dy2 = prefix[x3] - prefix[x2];
    return dx1 * dy2 < dy1 * dx2;
}

double slope(int i, int j, double* prefix) {
    return (prefix[j] - prefix[i]) / (j - i);
}

double findMaxAverage(int* nums, int numsSize, int k) {
    int n = numsSize;
    double* prefix = (double*)malloc((n + 1) * sizeof(double));
    prefix[0] = 0;
    for (int i = 0; i < n; ++i) {
        prefix[i + 1] = prefix[i] + nums[i];
    }

    int* dq = (int*)malloc((n + 1) * sizeof(int));
    int front = 0, back = 0;
    double res = -DBL_MAX;

    for (int j = k; j <= n; ++j) {
        int i = j - k;

        while (back - front >= 2 && is_concave(dq[back - 2], dq[back - 1], i, prefix)) {
            back--;
        }
        dq[back++] = i;

        while (back - front >= 2 && slope(dq[front], j, prefix) <= slope(dq[front + 1], j, prefix)) {
            front++;
        }

        res = max(res, slope(dq[front], j, prefix));
    }

    free(prefix);
    free(dq);
    return res;
}

File: leetcode/binary_search/644_Maximum_Average_Subarray_II/codes/convex_hull.py
from collections import deque
from itertools import accumulate
from math import inf
from typing import List

class Solution:
    def findMaxAverage(self, nums: List[int], k: int) -> float:
        n = len(nums)
        prefix = list(accumulate(nums, initial=0))  
        def slope(i: int, j: int) -> float:
            return (prefix[j] - prefix[i]) / (j - i)

        def is_concave(x1: int, x2: int, x3: int) -> bool:
            dx1, dy1 = x2 - x1, prefix[x2] - prefix[x1]
            dx2, dy2 = x3 - x2, prefix[x3] - prefix[x2]
            return dx1 * dy2 < dy1 * dx2

        res = -inf
        dq = deque()

        for j in range(k, n + 1):  
            i = j - k
            while len(dq) >= 2 and is_concave(dq[-2], dq[-1], i):
                dq.pop()
            dq.append(i)

            while len(dq) >= 2 and slope(dq[0], j) <= slope(dq[1], j):
                dq.popleft()

            res = max(res, slope(dq[0], j))

        return res
